File: articles/edgedetection/src/edge.cpp
// g++ edge.cpp -o edge `pkg-config --cflags --libs opencv`

#include <string>
#include <opencv2/opencv.hpp>
#include <opencv2/highgui.hpp>
#include <cassert>

int main(void)
{
	// load image
	std::string filename = "lena_color_512.png";
	cv::Mat image_orig = cv::imread(filename);
	if (image_orig.empty())
	{
		std::cout << "Cannot find or load image: " << filename << std::endl;
		return -1;
	}

	cv::Mat image = image_orig.clone();

	// convert image to grayscale
	if (image.channels() > 1)
		cv::cvtColor(image_orig, image, cv::COLOR_RGB2GRAY);

	// apply canny edge detector
	cv::Mat edges;
	cv::Canny(image, edges, 92, 128);

	// visualize image
	cv::imshow("image", image_orig);
	cv::imshow("edges", edges);
	cv::waitKey(0);

	cv::destroyAllWindows();

    // save image
    cv::imwrite("edges.png", edges);

	return 0;
}

File: articles/edgedetection/src/edge.py
import cv2

if __name__ == '__main__':

	# load image
	filename = '../../../shared/images/lena_color_512.tif'
	image_orig = cv2.imread(filename)
	if image_orig is None:
		print('Cannot find or load image:', filename)
		quit(-1)

	image = image_orig.copy()

	# convert image to grayscale
	if len(image.shape) > 2 and image.shape[-1] >= 3:
		image = cv2.cvtColor(image, cv2.COLOR_RGB2GRAY)

	# apply canny edge detector
	edges = cv2.Canny(image, 92, 128)

	# visualize image
	cv2.imshow('image', image_orig)
	cv2.imshow('edges', edges)
	cv2.waitKey(0)

	cv2.destroyAllWindows()
